### DISCUSSION TITLE
Interesting interaction between `set_up_ament=True` and nodes exiting gracefully

### DISCUSSION THREAD

**Discussion by cullenwren-volair** [mvukov/rules_ros2#168] [Q&A] [answered]
Was recently trying to figure out why I could not get my nodes to exit gracefully on shutdown and I traced it back to the `set_up_ament` flag being true in my nodes `ros_py_binary` rule. I was able to fail the test found in [examples/lifecycle](https://github.com/mvukov/rules_ros2/tree/main/examples/lifecycle) by setting `set_up_ament=True` for `lifecycle_talker` rule

Does this interaction make sense / is it intentional? 

To me, it seems to stem from that when `set_up_ament=True`, the PID of what `ros_launch` thinks the node is different from the nodes actual PID, but not sure why that would be.

Sure, steps to re-create:
in `examples/lifecycle/BUILD.bazel` make the change
```diff
ros2_cpp_binary(
    name = "lifecycle_talker",
    srcs = ["lifecycle_talker.cc"],
+   set_up_ament = True,
    deps = [
        "@ros2_common_interfaces//:cpp_std_msgs",
        "@ros2_rcl_interfaces//:cpp_lifecycle_msgs",
        "@ros2_rclcpp//:rclcpp",
        "@ros2_rclcpp//:rclcpp_lifecycle",
    ],
)
```
run the test in the `examples/lifecycle` folder
witrhin the log:
```
======================================================================
FAIL: test_talker_graceful_shutdown (tests.TestLifecyclePubSubAfterShutdown)
Test lifecycle talker graceful shutdown.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/9871e3f5514bc8b2cd22ccc79c1fe02d/sandbox/linux-sandbox/2/execroot/com_github_mvukov_rules_ros2_examples/bazel-out/k8-fastbuild/bin/lifecycle/tests.runfiles/com_github_mvukov_rules_ros2_examples/lifecycle/tests.py", line 149, in test_talker_graceful_shutdown
    launch_testing.asserts.assertExitCodes(proc_info, process=talker_node)
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/9871e3f5514bc8b2cd22ccc79c1fe02d/sandbox/linux-sandbox/2/execroot/com_github_mvukov_rules_ros2_examples/bazel-out/k8-fastbuild/bin/lifecycle/tests.runfiles/ros2_launch/launch_testing/launch_testing/asserts/assert_exit_codes.py", line 62, in assertExitCodes
    assert info.returncode in allowable_exit_codes, 'Proc {} exited with code {}'.format(
AssertionError: Proc lifecycle_talker-1 exited with code -15
``` 
full log: [test.log](https://github.com/mvukov/rules_ros2/files/12410010/test.log)

**Comment by mvukov** [comment]
Hi, can you paste the error log here for the reference, please?

**Reply by cullenwren-volair** [reply]
Added to the description

**Comment by ahans** [comment, marked as answer]
Great observation that it's due to the PID changing. When I read this, it was immediately clear that this is introduced by the wrapper that we use to set `AMENT_PREFIX_PATH`. I created #169 with a fix and that seems to make it work.

Since we use the same wrapper for `ros_py_binary`, this should fix that issue as well. @cullenwren-volair Could you give it a try please? Thanks!

**Reply by cullenwren-volair** [reply]
Those changes resolved it for me! Thanks for the quick response to this 😄
